File: python/GCD.py
# coding:utf8

# 求最大公约数

# 欧几里德定理：
# 定理： gcd(a, b) = gcd(b, a mod b)
# 证明：
#   a可以表示成a = kb + r, 则r = a mod b
#   假设d是a, b的一个公约数， 则有  d|a, d|b, 而r = a - kb, 因此d|r。
#   因此，d是(b, a mod b)的公约数。
#   加上d是(b，a mod b)的公约数，则d|b, d|r, 但是a = kb + r,因此d也是(a, b)的公约数。
#   因此，(a, b) 和(a, a mod b)的公约数是一样的，其最大公约数也必然相等，得证。

class Solution(object):
    def gcd(self, a, b):
        if b == 0:
            return a
        return self.gcd(b, a % b)


s = Solution()
print s.gcd(2, 5)
print s.gcd(2, 6)
print s.gcd(4, 10)

File: src/main/java/math/GCD.java
package math;

/**
 * The greatest common divisor (GCD) of two or more numbers is the greatest common factor
 * number that divides them, exactly.
 * 最大公约数
 */
public class GCD {
    public static void main(String[] args) {
        GCD gcd = new GCD();
        System.out.println(gcd.gcd(123456,7890));
    }
    /**
     * 返回最大公约数，基于欧几里得原理：gcd(a,b)=gcd(b,a mod b)。
     * 举例：
     * gcd(4,6)=gcd(4,2)=gcd(2,0)，最终返回2。
     * @param a
     * @param b
     * @return
     */
    public int gcd(int a,int b){
        if(a==0){
            return b;
        }
        if(b==0){
            return a;
        }
        int max=a,min=b;
        if(a<b){
            max=b;
            min=a;
        }
        return gcd(min,max%min);
    }
}
